lib/integrations/awsra/createsession/createsession.go

```diff
@@ -0,0 +1,301 @@
+	}
+
+	if req.PrivateKey == nil {
+		return trace.BadParameter("private key is required")
+	}
+
+	if req.httpClient == nil {
+		httpClient, err := defaults.HTTPClient()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		req.httpClient = httpClient
+	}
+
+	req.awsRegion = raTrustAnchor.Region
+	if err := aws.IsValidRegion(req.awsRegion); err != nil {
+		return trace.BadParameter("invalid region: %v", err)
+	}
+
+	req.clock = cmp.Or(req.clock, clockwork.NewRealClock())
+
+	return nil
+}
+
+// CreateSession exchanges a certificate for AWS credentials using the AWS IAM Roles Anywhere service.
+// This method is based on the following guide:
+// https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-sign-process.html
+func CreateSession(ctx context.Context, req CreateSessionRequest) (*CreateSessionResponse, error) {
+	if err := req.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Task 1: Create a canonical request
+	// https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-sign-process.html#authentication-task1
+
+	createSessionRequestBody := struct {
+		ProfileARN      string `json:"profileArn"`
+		RoleARN         string `json:"roleArn"`
+		TrustAnchorARN  string `json:"trustAnchorArn"`
+		RoleSessionName string `json:"roleSessionName,omitempty"`
+		DurationSeconds *int   `json:"durationSeconds,omitempty"`
+	}{
+		ProfileARN:      req.ProfileARN,
+		RoleARN:         req.RoleARN,
+		TrustAnchorARN:  req.TrustAnchorARN,
+		RoleSessionName: req.RoleSessionName,
+		DurationSeconds: req.DurationSeconds,
+	}
+
+	canonicalRequestBody, err := json.Marshal(createSessionRequestBody)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	canonicalRequestURL := &url.URL{
+		Scheme: "https",
+		Host:   fmt.Sprintf("rolesanywhere.%v.amazonaws.com", req.awsRegion),
+		Path:   "/sessions", // Task 1-2
+	}
+
+	canonicalRequest, err := http.NewRequestWithContext(ctx,
+		http.MethodPost, // Task 1-1
+		canonicalRequestURL.String(),
+		bytes.NewReader(canonicalRequestBody),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Task 1-4,5
+	formatedDate := req.clock.Now().UTC().Round(time.Second).Format("20060102T150405Z")
+	const signedHeaders = "content-type;host;x-amz-date;x-amz-x509"
+	canonicalRequest.Header.Set("Content-Type", "application/json")
+	canonicalRequest.Header.Set("Host", canonicalRequest.Host)
+	canonicalRequest.Header.Set("X-Amz-Date", formatedDate)
+	canonicalRequest.Header.Set("X-Amz-X509", base64.StdEncoding.EncodeToString(req.Certificate.Raw))
+
+	// Task 1-6
+	canonicalRequestBodyHash := sha256.Sum256(canonicalRequestBody)
+
+	// Task 1-7
+	canonicalRequestString := fmt.Sprintf(`POST
+/sessions
+
+content-type:application/json
+host:%s
+x-amz-date:%s
+x-amz-x509:%s
+
+%s
+%x`,
+		canonicalRequest.Header.Get("Host"),
+		canonicalRequest.Header.Get("X-Amz-Date"),
+		canonicalRequest.Header.Get("X-Amz-X509"),
+		signedHeaders,
+		canonicalRequestBodyHash,
+	)
```

Comment:
Manually crafting a string seems fragile, and I don't think we have guarantees that the canonical request will be built exactly like that. Can we call `canonicalRquest.Write` to recover the request content instead?